cvmfs/sync_union_tarball.cc

```diff
@@ -199,6 +164,74 @@ void SyncUnionTarball::Traverse() {
   }
 }
 
+void SyncUnionTarball::ProcessArchiveEntry(struct archive_entry *entry) {
+  std::string archive_file_path(archive_entry_pathname(entry));
+  std::string complete_path =
+      MakeCanonicalPath(base_directory_ + "/" + archive_file_path);
+
+  std::string parent_path;
+  std::string filename;
+  SplitPath(complete_path, &parent_path, &filename);
+
+  CreateDirectories(parent_path);
+
+  SharedPtr<SyncItem> sync_entry = SharedPtr<SyncItem>(new SyncItemTar(
+      parent_path, filename, src, entry, read_archive_signal_, this));
+
+  if (NULL != archive_entry_hardlink(entry)) {
+    const std::string hardlink =
+        base_directory_ + "/" + std::string(archive_entry_hardlink(entry));
+
+    printf("Hardlink %s -> %s\n", hardlink.c_str(), complete_path.c_str());
+
+    if (hardlinks_.find(hardlink) != hardlinks_.end()) {
+      hardlinks_.find(hardlink)->second.push_back(complete_path);
+    } else {
+      std::list<std::string> to_hardlink;
+      to_hardlink.push_back(complete_path);
+      hardlinks_[hardlink] = to_hardlink;
+    }
+    read_archive_signal_->Wakeup();
+    return;
+  }
+
+  if (sync_entry->IsDirectory()) {
+    if (know_directories_.find(complete_path) != know_directories_.end()) {
+      sync_entry->IsPlaceholderDirectory();
+    }
+    ProcessUnmaterializedDirectory(sync_entry);
+    dirs_[complete_path] = sync_entry;
+    know_directories_.insert(complete_path);
+
+    read_archive_signal_->Wakeup();  // We don't need to read data and we
+                                     // can read the next header
+
+  } else if (sync_entry->IsRegularFile()) {
```

Comment:
I think we should combine this case and the following case on, if the file is `.cvmfscatalog`, validate that we deal with a regular file.  Otherwise one could add a symlink .cvmfscatalog without magic behavior, which would be confusing.

Reply:
I am not quite sure I follow you on this one.

At the moment:
1. We check that it is a regular file
2. We process it
3. If is a catalog marker we keep track of it

A symlink to a catalog marker should not enter this branch at all, am I wrong?

Are you suggesting to don't process the symlink if it points to a catalog marker?